### ISSUE TITLE
Cache miss when transforming value with useMemo()

### ISSUE THREAD

**Issue by gonzochic** [vercel/swr#1357]
# Bug report

## Description / Observed Behavior
I am using dependant SWR requests. The first one is fetching an array of users, the second one is fetching additional information for each individual user. In order to reduce the amount of requests I filter the array of user for only active ones using `useMemo()`.

Although the resulting array of filtered user is exactly the same between page transition, it results in a cache miss (which means it doesn't return the cached value, instead waiting for the async request to complete).

## Expected Behavior

I expect SWR to return stale data first when I use exactly the same key + data input between page transitions.

## Repro Steps / Code Example
I created a Code Sandbox to reproduce the behaviour. 
[CodeSandbox](https://codesandbox.io/s/epic-haibt-b26sg?file=/src/App.js)

## Additional Context

My understanding is, that for dedouping we use `dequal`, which does a deep comparisons. I have found some information about how we evaluate, if the cache can provide stale data. The key is essential, but also the dependent data passed alongside the key is relevant. Here I expected also a deep comparison, but according to my example it doesn't seem to be the case.

Am I missing something?

SWR 0.5.6
React 17.0.2

**Comment by nataliemarleny** [comment]
Looking through the CodeSandbox example, these lines:

```
    if (users) return [users[0]];
```

and

```
    filteredUser ? ["/birthdays", filteredUser] : null,
```

cause the key to the second instantiation of `useSWR` to be `["/birthdays", ["alice"]]`

You can read why this is a problem here: [https://swr.vercel.app/docs/arguments#passing-objects](https://swr.vercel.app/docs/arguments#passing-objects)

**Comment by gonzochic** [comment]
Thanks for pointing me to the documentation, which means the behaviour is "as intended" :).

My question would be now, looking at my use case, if a shallow comparison is what the user is expecting (especially when a deep comparison is used for re-renders). I am also unsure if "my" case (having a complex object as input to another SWR), is really a common use-case, even if it feels that way for me.

I am afraid that the common solution, for these types of problems, could be using `JSON.stringify()` and `JSON.parse()` to compare the objects shallowly. From my perspective this would be an anti-pattern, but you know how things evolve ;).

 I created a new [Code Sandbox](https://codesandbox.io/s/goofy-lake-snjrn) showing what I mean.

**Comment by nataliemarleny** [comment]
@gonzochic Using `JSON.stringify` and `JSON.parse` seems like a reasonable approach to me with some caveats:

1. the object should be serializable of course,
2. `JSON.stringify` doesn't guarantee the order of the keys in the string, so it could be better to use something like `json-stable-stringify`.

I've looked through the docs to see if there's a general answer for how to handle this, but I haven't found one. I remember reading in some issues that the recommended approach is to spread the nested array into the array that represents the `useSWR` key. There are limitations to this approach as well of course.

**Comment by gonzochic** [comment]
Thanks for your answer! Is this something that we want to have in SWR or do we want to leave it to the user of the library? Since `dequal` is already in the dependeny tree this might only be a code adaption. I could look into it, but just want to be sure if this is in-line with the product vision.

**Comment by shuding** [comment]
As a side note, you can also use the Middleware feature in 1.0 to add key serialization: https://swr.vercel.app/docs/middleware#serialize-object-keys

We're considering adding deep comparison/serialization for keys into SWR, but some benchmarking work needs to be done first. In worst cases there can be a huge amount of components rendered with `useSWR` inside, and they can use large objects as keys. It would be great to measure the performance impact first 👍

**Comment by shuding** [comment]
Closing since key serialization was added in #1429 and released as 1.1.0-beta.0, you don't need `useMemo` anymore! 🎉
